google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,353 @@
+
+  /** Whether a query stream is currently in progress. Only one stream can be run at a time. */
+  private boolean streamInProgress = false;
+
+  /** Query limit to use when fetching all descendants. */
+  private int maxPendingOps = MAX_PENDING_OPS;
+
+  /**
+   * The number of pending BulkWriter operations at which RecursiveDelete starts the next limit
+   * query to fetch descendants.
+   */
+  private int minPendingOps = MIN_PENDING_OPS;
+
+  /**
+   * The last document snapshot returned by the stream. Used to set the startAfter() field in the
+   * subsequent stream.
+   */
+  @Nullable private DocumentSnapshot lastDocumentSnap;
+
+  /**
+   * The number of pending BulkWriter operations. Used to determine when the next query can be run.
+   */
+  private int pendingOperationsCount = 0;
+
+  RecursiveDelete(
+      FirestoreRpcContext<?> firestoreRpcContext,
+      BulkWriter writer,
+      CollectionReference reference) {
+    Preconditions.checkState(reference != null, "CollectionReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.collectionReference = reference;
+    this.documentReference = null;
+  }
+
+  RecursiveDelete(
+      FirestoreImpl firestoreRpcContext, BulkWriter writer, DocumentReference reference) {
+    Preconditions.checkState(reference != null, "DocumentReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.documentReference = reference;
+    this.collectionReference = null;
+  }
+
+  public ApiFuture<Void> run() {
+    Preconditions.checkState(
+        documentsPending, "The recursive delete operation has already been completed");
+
+    // TODO: figure out if this is safe, or a way to lock this.
+    writer.verifyNotClosedLocked();
+
+    streamDescendants();
+    return completionFuture;
+  }
+
+  private void streamDescendants() {
+    Preconditions.checkState(
+        (documentReference == null && collectionReference != null)
+            || (documentReference != null && collectionReference == null),
+        "Either documentReference or collectionReference should be set");
+
+    Query query;
+    if (collectionReference != null) {
+      query = getAllDescendantsQuery(collectionReference);
+    } else {
+      query = getAllDescendantsQuery(documentReference);
+    }
+    streamInProgress = true;
+    final int[] streamedDocsCount = {0};
+    final ApiStreamObserver<DocumentSnapshot> responseObserver =
+        new ApiStreamObserver<DocumentSnapshot>() {
+          public void onNext(DocumentSnapshot snapshot) {
+            streamedDocsCount[0]++;
+            lastDocumentSnap = snapshot;
+            deleteReference(snapshot.getReference());
+          }
+
+          public void onError(Throwable throwable) {
+            lastError =
+                new FirestoreException("Failed to fetch children documents.", Status.UNAVAILABLE);
```

Comment:
This swallows the underlying exception.

Reply:
Wrapped the old exception. The status should be `UNAVAILABLE` regardless of the error, as it is on Node.